aws-sdk-s3.go

```diff
@@ -0,0 +1,49 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func GetS3Client() *s3.S3 {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+
+	svc := s3.New(sess)
+
+	return svc
+}
+
+/*
+	S3ServiceInterface created for the purpose of testing
+	When testing AwsSdkCountObjectsInBucket, I need to mock the `svc` parameter
+	If I defined `svc` of type *s3.S3, all *s3.S3 functions need to be mocked
+	By creating an interface, I can only define functions I use
+	Thus, only the functions I use have to be mocked
+	And an argument of type *s3.S3 would satisfy the interface
+*/
```

Comment:
Is there a Go equivalent to mocking the import and then providing return values when the import's public functions are used? Similar to how one would it in python:

```python
import boto3

@patch('somepath.boto3')
def test_my_func(mock_boto3):
    # Arrange
    mock_boto3.client.S3.list_objects_v2.return_value = {'Name':'my_object'}
    # Act
    result = my_func(mock_boto3)
    # Assert
    assert 1 == result
````

Go seems to be:
1. Make a struct
2. Define necessary functions so that the struct can satisfy the interface that is param to the function (where the functions return the mocked values I want)
3. Create an object of type struct and pass it in

Reply:
there are a couple testing/mocking frameworks that give something like this.

pretty sure https://github.com/golang/mock is what you are looking for. 

unlike python, it requires some code generation to work.